src/graphql/ObjectTransformer.js

```diff
@@ -0,0 +1,363 @@
+/*******************************************************************************
+ *
+ *    Copyright 2018 Adobe. All rights reserved.
+ *    This file is licensed to you under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License. You may obtain a copy
+ *    of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software distributed under
+ *    the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+ *    OF ANY KIND, either express or implied. See the License for the specific language
+ *    governing permissions and limitations under the License.
+ *
+ ******************************************************************************/
+
+'use strict';
+
+const recursiveMerge = require('./utils').recursiveMerge;
+
+/**
+ * An ObjectTransformer transforms objects according to their transform rules:
+ * 
+ * adders on objects will add all the fields in adder.add if adder.when is present (to the object itself or subObjects)
+ * (it will add all the fields in adder.add in anycase if adder.when is not specified)
+ * 
+ * ignore on objects will remove all the ignore fields if present (of the object itself)
+ * 
+ * moveFields on objects will move all the mover.fields (or all the present fields if mover.fileds is not specified)
+ * from mover.from (or from the object itself if mover.from is not specified) to the mover.to field
+ * 
+ * alias on objects will set the property '__aliasFor' of this object equal to it's value
+ * 
+ * args on objects will add the args fields to the '__args' property of the object
+ * 
+ * inlineFragments adds inline fragments to the object if any of the inline fragment field is present
+ */
+class ObjectTransformer {
+
+    /**
+     * 
+     * @param {object} transformRules     object represantation of graphql schema containing the transforms
+     */
+    constructor(transformRules) {
+        this.transformRules = transformRules || {};
+    }
+
+    /**
+     * transforms/modifies the toTransform object according to the transformRules
+     * 
+     * 1. adders
+     * 2. transforms on subobjects
+     * 3. args
+     * 4. ignore
+     * 5. move
+     * 6. alias
+     * 7. inlineFragments
+     * 
+     * @param {object} toTransform         object to be transformed
+     * @param {object} transformRules      object with all the relevant transforms for the current object
+     * 
+     * @returns {boolean} indicating if the object results in an empty object (should never be the case for main query)
+     */
+    transform(toTransform, transformRules) {
```

Comment:
Do we really need the `transformRules` parameter here since this is passed to the constructor? Is there a need to be able to override the rules passed to the constructor? If not, please remove.

Reply:
yes, it's a recursive function, only the very first time it will be equal to the transformRules of the constructor

Reply:
A better design would be to only expose `transform(toTransform)` which calls a private `_transform(toTransform, transformRules)` method.